### ISSUE TITLE
`3.0.0rc1` slow target generation

### ISSUE THREAD

**Issue by Vaccummer** [DeepLabCut/DeepLabCut#2632]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Bug description

I found in PoseTrainingRunner.step(), this line "target = underlying_model.get_target(outputs, batch["annotations"])" consumes most time of a step, which is much more than model.forward and loss.bakwards.
I test time consumption of every part of a training step, and the result is :
![image](https://github.com/DeepLabCut/DeepLabCut/assets/129358302/2df740da-f1ee-419e-a87e-d432caa9cf44)

                

### Operating System

Ubuntu20.04

### DeepLabCut version

dlc_3.0rc

### DeepLabCut mode

single animal

### Device type

GPU-Nvidia V100 smx2

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/DeepLabCut/DeepLabCut/blob/master/CODE_OF_CONDUCT.md)

**Comment by Vaccummer** [comment]
![image](https://github.com/DeepLabCut/DeepLabCut/assets/129358302/8b937b7a-21be-48fa-8ecd-8d0d1bf6ed9d)

.get_target() is defined here
but I don't exactly know what this function's effect, I guess it's to process backbone output.

**Comment by Vaccummer** [comment]
this is my pytorch_config.yaml

```
data:
  colormode: RGB
  inference:
    normalize_images: true
    auto_padding:
      pad_width_divisor: 32
      pad_height_divisor: 32
  train:
    affine:
      p: 0.5
      rotation: 30
      scaling:
      - 1.0
      - 1.0
      translation: 0
    collate:
      type: ResizeFromDataSizeCollate
      min_scale: 0.4
      max_scale: 1.0
      min_short_side: 128
      max_short_side: 1152
      multiple_of: 32
      to_square: false
    covering: false
    gaussian_noise: 12.75
    hist_eq: false
    motion_blur: false
    normalize_images: true
device: auto
metadata:
  project_path: /home/am/dlc/Vgat/24.05.29-G80-hrnet
  pose_config_path: 
    /home/am/dlc/Vgat/24.05.29-G80-hrnet/dlc-models-pytorch/iteration-0/24.05.29Jun6-trainset95shuffle1/train/pose_cfg.yaml
  bodyparts:
  - pupilM
  - pupilL
  - pupilU
  - pupilD
  - pupilLU
  - pupilRD
  - pupilRU
  - pupilLD
  - eyeM
  - eyeL
  - eyeU
  - eyeD
  - earD
  - earM
  - earU
  - earL
  - earR
  - noseU
  - noseD
  - noseMU
  - noseL
  - noseMD
  - mouthM
  - mouthR
  - mouthL
  - whisker1O
  - whisker1M
  - whisker1L
  - whisker1E
  - whisker2O
  - whisker2M
  - whisker2L
  - whisker2E
  - whisker3O
  - whisker3M
  - whisker3L
  - whisker3E
  unique_bodyparts: []
  individuals:
  - animal
  with_identity:
method: bu
model:
  backbone:
    type: HRNet
    model_name: hrnet_w48
    freeze_bn_stats: false
    freeze_bn_weights: false
    interpolate_branches: false
    increased_channel_count: false
  backbone_output_channels: 48
  heads:
    bodypart:
      type: HeatmapHead
      weight_init: normal
      predictor:
        type: HeatmapPredictor
        apply_sigmoid: false
        clip_scores: true
        location_refinement: true
        locref_std: 7.2801
      target_generator:
        type: HeatmapGaussianGenerator
        num_heatmaps: 37
        pos_dist_thresh: 17
        heatmap_mode: KEYPOINT
        generate_locref: true
        locref_std: 7.2801
      criterion:
        heatmap:
          type: WeightedMSECriterion
          weight: 1.0
        locref:
          type: WeightedHuberCriterion
          weight: 0.05
      heatmap_config:
        channels:
        - 48
        - 37
        kernel_size:
        - 3
        strides:
        - 2
      locref_config:
        channels:
        - 48
        - 74
        kernel_size:
        - 3
        strides:
        - 2
net_type: hrnet_w48
runner:
  type: PoseTrainingRunner
  gpus:
  key_metric: test.mAP
  key_metric_asc: true
  eval_interval: 1
  optimizer:
    type: AdamW
    params:
      lr: 0.0001
  scheduler:
    type: LRListScheduler
    params:
      lr_list:
      - - 1e-05
      - - 1e-06
      milestones:
      - 160
      - 190
  snapshots:
    max_snapshots: 50
    save_epochs: 10
    save_optimizer_state: false
train_settings:
  batch_size: 2
  dataloader_workers: 4
  dataloader_pin_memory: true
  display_iters: 500
  epochs: 500
  seed: 42
mixed_precision: false
```

**Comment by n-poulsen** [comment]
Hi @Vaccummer! Thanks for reporting this, I'll look into it and reproducing this and get back to you

**Comment by Vaccummer** [comment]
```
        for b in range(batch_size):
            for heatmap_idx, group_keypoints in enumerate(coords[b]):
                for keypoint in group_keypoints:
                    # FIXME: Gradient masking weights should be parameters
                    if keypoint[-1] == 0:
                        # full gradient masking
                        weights[b, heatmap_idx] = 0.0
                    elif keypoint[-1] == -1:
                        # full gradient masking
                        weights[b, heatmap_idx] = 0.0
                    elif keypoint[-1] > 0:
                        # keypoint visible
                        a += 1
                        self.update(
                            heatmap=heatmap[b, :, :, heatmap_idx],
                            grid=grid,
                            keypoint=keypoint[..., :2],
                            locref_map=self.get_locref(locref_map, b, heatmap_idx),
                            locref_mask=self.get_locref(locref_mask, b, heatmap_idx),
                        )
```
I found the reason , in HeatmapGenerator.forward(), this cycle will conduct self.update for (point_num*batch_size) times in a step, without any parallel method. I has 37 point to track and batch_size=2, so this function will conduct 84 times in a step, which takes a long time.

**Comment by roaldarbol** [comment]
Maybe an associated issue. I just trained a model, and began to analyze a video, but am only getting tracking speed of around 1.33fps (note the videos are really long, over 500.000 frames) - if I remember correctly, I used to get speeds quite a lot faster than that for similar videos.
```python
INFO:GUI:Videos selected to analyze:
{'E:\\004-Trackball-Breathing\\Data\\InputData\\VideosMouthparts\\front2024-02-07T14_18_46.avi'}
INFO:GUI:Filtering predictions ENABLED
INFO:GUI:Plot trajectories ENABLED.
INFO:console:Analyzing videos with C:\Users\mr630\Desktop\DLC-Mouthparts\DLC-Project\dlc-models-pytorch\iteration-0\BeetleMouthpartsJun25-trainset95shuffle1\train\snapshot-200.pt
  0%|                                                                                                                                                            | 0/553717 [00:00<?, ?it/s]
INFO:console:Starting to analyze E:\004-Trackball-Breathing\Data\InputData\VideosMouthparts\front2024-02-07T14_18_46.avi
INFO:console:Video metadata:
  Overall # of frames:    553717
  Duration of video [s]:  9228.62
  fps:                    60.0
  resolution:             w=1280, h=720

INFO:console:Running Pose Prediction
  0%|                                                                                                                                              | 149/553717 [03:19<207:14:23,  1.35s/it]
```

Here's the `config.yaml`:
```yaml
# Project definitions (do not edit)
Task: BeetleMouthparts
scorer: roaldarboel
date: Jun25
multianimalproject: false
identity:


# Project path (change when moving around)
project_path: C:\Users\mr630\Desktop\DLC-Mouthparts\DLC-Project


# Default DeepLabCut engine to use for shuffle creation (either pytorch or tensorflow)
engine: pytorch


# Annotation data set configuration (and individual video cropping parameters)
video_sets:
  C:\Users\mr630\Desktop\DLC-Mouthparts\DLC-Project\videos\front2024-01-15T14_31_22.avi:
    crop: 0, 1280, 0, 720
bodyparts:
- eye_L
- eye_R
- cavity_center
- maxilla_mid_L
- maxilla_end_L
- maxilla_mid_R
- maxilla_end_R
- labium_mid_L
- labium_end_L
- labium_mid_R
- labium_end_R


# Fraction of video to start/stop when extracting frames for labeling/refinement
start: 0
stop: 1
numframes2pick: 20


# Plotting configuration
skeleton:
- [cavity_center, eye_L]
- [cavity_center, eye_R]
- [cavity_center, maxilla_mid_L]
- [cavity_center, maxilla_mid_R]
- [cavity_center, labium_mid_L]
- [cavity_center, labium_end_R]
- [maxilla_mid_L, maxilla_end_L]
- [maxilla_mid_R, maxilla_end_R]
- [labium_mid_L, labium_end_L]
- [labium_mid_R, labium_end_R]
skeleton_color: black
pcutoff: 0.6
dotsize: 12
alphavalue: 0.7
colormap: rainbow


# Training,Evaluation and Analysis configuration
TrainingFraction:
- 0.95
iteration: 0
default_net_type: resnet_50
default_augmenter: default
snapshotindex: -1
detector_snapshotindex: -1
batch_size: 8


# Cropping Parameters (for analysis and outlier frame detection)
cropping: false
#if cropping is true for analysis, then set the values here:
x1: 0
x2: 640
y1: 277
y2: 624


# Refinement configuration (parameters from annotation dataset configuration also relevant in this stage)
corner2move2:
- 50
- 50
move2corner: true


# Conversion tables to fine-tune SuperAnimal weights
SuperAnimalConversionTables:
```

**Comment by roaldarbol** [comment]
EDIT: It seems that the GPU is not getting activated for tracking:
![image](https://github.com/DeepLabCut/DeepLabCut/assets/25629697/6a7c889e-714f-4bb7-94de-0efac7808c17)

I'm quite sure it was engaged during training, but I'll do a bit more testing on Thursday.

**Comment by roaldarbol** [comment]
On my end, it was a CUDA/Pytorch problem, related to installing with pixi. Went from 1.33fps to 7.3fps with CUDA engaged.

**Comment by Vaccummer** [comment]
I find out the resolution, use cupy instead of numpy in the HeatmapGenerator.update() can dramatically shrink the time consumption of head.tatget_generate(), which comes to 0.06s compared to 0.8s of numpy.(Remember to delete cupy object or you will encounter CUDA OOM.

**Comment by roaldarbol** [comment]
@Vaccummer Does this also affect the `Analyze videos` step, or is it restricted to the training?

**Comment by Vaccummer** [comment]
> @Vaccummer Does this also affect the `Analyze videos` step, or is it restricted to the training?

no effect, this only works in train network.

**Comment by roaldarbol** [comment]
@Vaccummer Fair enough - still a good improvement it seems! Have you considered opening a PR with your `cupy` solution? Might speed up the work. :-)

**Comment by n-poulsen** [comment]
@Vaccummer thanks for working on this! We haven't found slow target generation like this during our testing, but will look to improve performance there as well. If you want to open a PR with your solution we'll look into it :)

@roaldarbol which model are you running inference with?

**Comment by Vaccummer** [comment]
> @Vaccummer thanks for working on this! We haven't found slow target generation like this during our testing, but will look to improve performance there as well. If you want to open a PR with your solution we'll look into it :)
> 
> @roaldarbol which model are you running inference with?

only in this occasion where you choose hrnet58 and have too many dots to track, you will encounter  target_generation performance problem. Resnet doesn't have this problem.  Use cupy instead of numpy in HeatmapGenerator.update() will accelerate this process, but 1-2G of GPU memory will be allocated.

**Comment by n-poulsen** [comment]
Closing as stale
